### PR TITLE
3732 costmap raw update msgs

### PR DESCRIPTION
**Pull request by emSko**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3732  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo simulation |

---

## Description of contribution in a few bullet points

* Adding new msg type **CostmapUpdate.msg**,
* Adding **CostmapUpdate** publisher to  **Costmap2DPublisher** class,
* Re-implementation of the **CostmapSubscriber** class to support the **CostmapUpdate** msg. 

## Description of documentation updates required from your changes

---

## Future work that may be required in bullet points

* I see a lot of redundancy in **Costmap2DPublisher** class while creating the msgs, adding methods that are responsible for populating msgs and return the unique ptr to msgs might improve the code readability

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp [hunk 1]
```diff
@@ -101,6 +103,7 @@ class Costmap2DPublisher
     costmap_pub_->on_deactivate();
     costmap_update_pub_->on_deactivate();
     costmap_raw_pub_->on_deactivate();
+    costmap_raw_update_pub_->on_activate();
```
Comment (harderthan) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp:106]:
```suggestion
    costmap_raw_update_pub_->on_deactivate();
```

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp [hunk 2]
```diff
@@ -136,6 +139,8 @@ class Costmap2DPublisher
   void prepareGrid();
   void prepareCostmap();
 
+  std::unique_ptr<nav2_msgs::msg::CostmapUpdate> get_raw_costmap_update_msg();
```
Comment (harderthan) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp:142]: It is better to use Camel Case, as shown below.  

```suggestion
  std::unique_ptr<nav2_msgs::msg::CostmapUpdate> getRawCostmapUpdateMsg();
```

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 3]
```diff
@@ -26,72 +27,130 @@ CostmapSubscriber::CostmapSubscriber(
...
 {
   auto node = parent.lock();
+  logger_ = node->get_logger();
   costmap_sub_ = node->create_subscription<nav2_msgs::msg::Costmap>(
     topic_name_,
     rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
     std::bind(&CostmapSubscriber::costmapCallback, this, std::placeholders::_1));
+  costmap_update_sub_ = node->create_subscription<nav2_msgs::msg::CostmapUpdate>(
+    topic_name_ + "_updates",
+    rclcpp::QoS(rclcpp::KeepLast(10)).transient_local().reliable(),
```
Comment (harderthan) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:37]: I was just wondering if you could kindly explain why a queue size of '10' was chosen?

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 4]
```diff
@@ -26,72 +27,130 @@ CostmapSubscriber::CostmapSubscriber(
...
+    std::bind(&CostmapSubscriber::costmapUpdateCallback, this, std::placeholders::_1));
 }
 
+
```
Comment (harderthan) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:41]: There are two lines.  

```suggestion
```

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 5]
```diff
@@ -181,13 +183,30 @@ void Costmap2DPublisher::prepareCostmap()
   }
 }
 
-void Costmap2DPublisher::publishCostmap()
+std::unique_ptr<nav2_msgs::msg::CostmapUpdate> Costmap2DPublisher::get_raw_costmap_update_msg()
```
Comment (harderthan) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp:186]: I was just wondering if you could kindly explain why you use `std::unique_ptr` for the return type.   
Returning the variable with `std::unique_ptr` is unnecessary due to Return Value Optimization (RVO).

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 3 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: Updates are generated only for the parts of the map where the change has occurred (new cleaning or marking observation). It might happen that some changes will be available only in a single update msg. So I guess that my intuition was to extend the queue size to prevent losing this information when the subscriber is waiting for the collision checker or other clients to end their computations. 
Size 10 was just arbitrarily selected as a default value for QoS in ROS 2. I am open to suggestions and discussion if the queue is even necessary here.

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 6]
```diff
@@ -220,6 +243,9 @@ void Costmap2DPublisher::publishCostmap()
       }
       costmap_update_pub_->publish(std::move(update));
     }
+    if (costmap_raw_update_pub_->get_subscription_count() > 0) {
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp:246]: Can you break the occupancy grid update population in the lines above this into its own function similar to createRawCostmapUpdateMsg called `createGridUpdateMsg`

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 7]
```diff
@@ -181,13 +183,30 @@ void Costmap2DPublisher::prepareCostmap()
   }
 }
 
-void Costmap2DPublisher::publishCostmap()
+std::unique_ptr<nav2_msgs::msg::CostmapUpdate> Costmap2DPublisher::createRawCostmapUpdateMsg()
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp:186]: Rename `createCostmapUpdateMsg`

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 8]
```diff
@@ -14,84 +14,158 @@
...
+    costmap_received_ = true;
   }
+}
+
+void CostmapSubscriber::costmapUpdateCallback(
+  const nav2_msgs::msg::CostmapUpdate::SharedPtr update_msg)
+{
+  if (costmap_received_) {
+    if (costmap_msg_ != nullptr) {
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:95]:
```suggestion
    if (costmap_msg_) {
```

Is the same :-)

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 9]
```diff
@@ -14,84 +14,158 @@
...
+      map_call_size_y < update_msg->y + update_msg->size_y)
+    {
+      RCLCPP_WARN(
+        logger_, "Update area outside of original map area. Costmap bounds: %d X %d, "
+        "Update origin: %d, %d  bounds: %d X %d", map_cell_size_x, map_call_size_y,
+        update_msg->x, update_msg->y, update_msg->size_x, update_msg->size_y);
+      return;
     }
+
+    // todo: check if frame is the same
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:114]: You can assume its in the right frame

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 10]
```diff
@@ -14,84 +14,158 @@
...
+
+    auto master_array = costmap_->getCharMap();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:116]: Make sure this is explicitly a reference, since if its a copy, it won't update the master grid

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 11]
```diff
@@ -14,84 +14,158 @@
...
+  std::scoped_lock lock(*(costmap_->getMutex()), costmap_msg_mutex_);
+  if (haveCostmapParametersChanged()) {
+    costmap_->resizeMap(
+      costmap_msg_->metadata.size_x, costmap_msg_->metadata.size_y,
+      costmap_msg_->metadata.resolution,
+      costmap_msg_->metadata.origin.position.x,
+      costmap_msg_->metadata.origin.position.y);
   }
+
+  auto master_array = costmap_->getCharMap();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:142]: Ditto

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Overall, very few notes, well done!

This does require some unit testing to make sure it works as expected :smile:

File: nav2_msgs/msg/CostmapUpdate.msg [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# Update msg for Costmap containing the modified part of Costmap
+std_msgs/Header header
+
+uint32 x
+uint32 y
+
+uint32 size_x
```
[the remark this answers is not in the document]
Comment (emSko) [reply nav2_msgs/msg/CostmapUpdate.msg:7]: In my opinion, the names `width` and `height` are a bit confusing, since they do not clearly indicate the connection to the respective dimensions. Also, I took the naming convention from `CostmapMetaData.msg`

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp (hunk 6 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp:246]: As agreed at the last working group meeting this work is out of scope of this issue. I have created a separate issue for that: #3956 3956

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp (hunk 5 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: Publishing the `std::unique_ptr` instead of raw msg is a principle adopted in this project, so that is why I've decided to create msg in unique_ptr directly.
[Change publishers to publish unique ptrs ](https://github.com/ros-planning/navigation2/pull/1673)

I believe that the cause for that is related with `Publisher` class implementation. 
When calling `publish(const T & msg)` additional copy of msg is made when using the intra-process communication
[publisher.hpp:L278](https://github.com/ros2/rclcpp/blob/9c098e544ecf191b7c61f63f7f4fac6f6449cedd/rclcpp/include/rclcpp/publisher.hpp#L278), and in the end you have a copy stored in the unique_ptr and pass the ownership further.
When using ` publish(std::unique_ptr<MessageT, MessageDeleter> msg)` you are omitting this copy. Also please take a look at the comments describing each of the overloaded methods.

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 9 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: Also in general, we don't allow TODOs in line of the code. If you have some questions - please ask :smile:

**Comment by SteveMacenski** [comment]
@emSko please don't hit "resolve" on things that haven't been pushed / completed yet. I went through and unresolved a number of items that haven't been addressed yet. It helps with my tracking of status :smile:

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 8 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: resolved with: 33bf135588a90c6db0f4745b1155de7e00941f2c

Please resolve this conversation if you can see these changes now.

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 9 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: resolved with: 33bf135588a90c6db0f4745b1155de7e00941f2c

Please resolve this conversation if you can see these changes now.

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 10 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: resolved with: 33bf135588a90c6db0f4745b1155de7e00941f2c

Please resolve this conversation if you can see these changes now.

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 11 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: resolved with: 33bf135588a90c6db0f4745b1155de7e00941f2c

Please resolve this conversation if you can see these changes now.

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 13]
```diff
@@ -220,6 +243,9 @@ void Costmap2DPublisher::publishCostmap()
       }
       costmap_update_pub_->publish(std::move(update));
     }
+    if (costmap_raw_update_pub_->get_subscription_count() > 0) {
+      costmap_raw_update_pub_->publish(createRawCostmapUpdateMsg());
```
[the remark this answers is not in the document]
Comment (emSko) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: In this case, I believe that using `std::move() `explicitly might cause an unnecessary call to the move constructor of `unique_ptr`. 
I have run a toy example which showed that due to RVO/NRVO optimization, no additional call to the constructor was made. However, when calling std::move() I triggered the move constructor.
When I turned off the optimization I also had a call to move constructor. 

So to sum up, we are ending up with the move constructor call in the worst-case scenario, but when calling move, it is more likely, because it seems like it blocks RVO.

**Comment by SteveMacenski** [comment]
Try rebasing and pushing a new PR with a new branch. Something is not quite right here the fact that some of your commits are not appearing on github's diff UX. Everytime that's happened in the past and I overlooked it as a quirk of github messing something up, there ended up being real issues that bit us in the butt later.

**Comment by emSko** [comment]
@SteveMacenski I have opened a new PR  with rebased changes as you asked: #3965 , so hopefully the issues with Github are resolved now.
Shall we close this PR and move with the review to the new one?
I am still working on tests so there are going to be a couple more commits.

**Comment by SteveMacenski** [comment]
Yes, sounds good! I see things properly now in the new PR

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp (hunk 6 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp:246]: I included that suggestion in the scope of the issue related to this PR and its continuation.
Resolved with bad53b5a122b1ae1af254de3298d233f9972e0a7
